homeassistant/components/google_mail/sensor.py

```diff
@@ -0,0 +1,156 @@
+
+from .const import (
+    ATTR_ENABLED,
+    ATTR_END,
+    ATTR_MESSAGE,
+    ATTR_PLAIN_TEXT,
+    ATTR_RESTRICT_CONTACTS,
+    ATTR_RESTRICT_DOMAIN,
+    ATTR_START,
+    ATTR_TITLE,
+    DOMAIN,
+)
+from .entity import GoogleMailEntity
+
+SCAN_INTERVAL = timedelta(minutes=15)
+
+SERVICE_SET_VACATION = "set_vacation"
+
+SERVICE_SCHEMA = vol.All(
+    cv.make_entity_service_schema(
+        {
+            vol.Required(ATTR_ENABLED, default=True): cv.boolean,
+            vol.Optional(ATTR_TITLE): cv.string,
+            vol.Required(ATTR_MESSAGE): cv.string,
+            vol.Optional(ATTR_PLAIN_TEXT, default=True): cv.boolean,
+            vol.Optional(ATTR_RESTRICT_CONTACTS): cv.boolean,
+            vol.Optional(ATTR_RESTRICT_DOMAIN): cv.boolean,
+            vol.Optional(ATTR_START): cv.date,
+            vol.Optional(ATTR_END): cv.date,
+        },
+    )
+)
+
+SENSOR_TYPE = SensorEntityDescription(
+    key="vacation_end_date",
+    name="Vacation end date",
+    icon="mdi:clock",
+    device_class=SensorDeviceClass.TIMESTAMP,
+)
+
+
+async def async_setup_service(hass: HomeAssistant) -> None:
+    """Add the services for Google Mail."""
+
+    async def extract_gmail_config_entries(call: ServiceCall) -> list[ConfigEntry]:
+        return [
+            entry
+            for entry_id in await async_extract_config_entry_ids(hass, call)
+            if (entry := hass.config_entries.async_get_entry(entry_id))
+            and entry.domain == DOMAIN
+        ]
+
+    def _set_vacation(call: ServiceCall, entry: ConfigEntry) -> None:
+        """Run vacation call in the executor."""
+        credentials = Credentials(entry.data[CONF_TOKEN][CONF_ACCESS_TOKEN])
+        service = build("gmail", "v1", credentials=credentials)
+        settings = {
+            "enableAutoReply": call.data[ATTR_ENABLED],
+            "responseSubject": call.data.get(ATTR_TITLE),
+        }
+        if contacts := call.data.get(ATTR_RESTRICT_CONTACTS):
+            settings["restrictToContacts"] = contacts
+        if domain := call.data.get(ATTR_RESTRICT_DOMAIN):
+            settings["restrictToDomain"] = domain
+        if _date := call.data.get(ATTR_START):
+            _dt = datetime.combine(_date, datetime.min.time())
+            settings["startTime"] = _dt.timestamp() * 1000
+        if _date := call.data.get(ATTR_END):
+            _dt = datetime.combine(_date, datetime.min.time())
+            settings["endTime"] = (_dt + timedelta(days=1)).timestamp() * 1000
+        if call.data[ATTR_PLAIN_TEXT]:
+            settings["responseBodyPlainText"] = call.data[ATTR_MESSAGE]
+        else:
+            settings["responseBodyHtml"] = call.data[ATTR_MESSAGE]
+        _settings = service.users().settings()  # pylint: disable=no-member
+        try:
+            _settings.updateVacation(userId="me", body=settings).execute()
+        except RefreshError as ex:
+            entry.async_start_reauth(hass)
```

Comment:
I believe the threadsafe error is because this async_start_reauth has to be called from the async thread context, but this is being called in a background thread.

I think you may need to move this try except block until `set_vacation` so that its thrown from the sync context back into the async context.

Reply:
When doing that, you can also change this function to accept an access token or Credentials object rather than the config entry since it will no longer need the full entry.